File: src/engine/validation.ts
import { RawEvent, ValidationError, SourceType } from '../types.js';

export const VALID_SOURCES = new Set<string>([
  'organic_search',
  'paid_search',
  'social',
  'email',
  'website',
  'manual',
]);

export const VALID_EVENT_TYPES = new Set([
  'page_visit',
  'campaign_click',
  'form_submission',
  'email_open',
  'email_click',
  'contacted',
  'qualified',
  'converted',
  'lost',
]);

export function isValidIsoTimestamp(str: any): boolean {
  if (!str || typeof str !== 'string') return false;
  // Check basic ISO format or Date parseability
  const parsed = Date.parse(str);
  if (isNaN(parsed)) return false;
  // Require reasonable year and structure (e.g. 2020-2035)
  const d = new Date(parsed);
  return d.getFullYear() >= 1990 && d.getFullYear() <= 2100;
}

export function validateEvent(raw: RawEvent): {
  isValid: boolean;
  errors: ValidationError[];
} {
  const errors: ValidationError[] = [];

  if (!raw || typeof raw !== 'object') {
    return {
      isValid: false,
      errors: [
        {
          event_id: 'unknown',
          error_type: 'MALFORMED_RECORD',
          field: 'record',
          message: 'Event record must be a valid JSON object or CSV row.',
          raw_record: raw,
        },
      ],
    };
  }

  // 1. Validate event_id
  const eventId = raw.event_id ? String(raw.event_id).trim() : '';
  if (!eventId) {
    errors.push({
      event_id: 'missing-id',
      error_type: 'MISSING_EVENT_ID',
      field: 'event_id',
      message: 'Event record is missing a required event_id.',
      raw_record: raw,
    });
  }

  // 2. Validate lead identification (at least one of email, phone, or explicit lead_id)
  const email = raw.email ? String(raw.email).trim() : '';
  const phone = raw.phone ? String(raw.phone).trim() : '';
  const leadId = raw.lead_id ? String(raw.lead_id).trim() : '';

  if (!email && !phone && !leadId) {
    errors.push({
      event_id: eventId || 'missing-id',
      error_type: 'MISSING_LEAD_IDENTITY',
      field: 'email/phone/lead_id',
      message: 'Event must contain at least one usable identifier (email, phone, or lead_id).',
      raw_record: raw,
    });
  }

  // 3. Validate timestamp
  const timestamp = raw.timestamp;
  if (!isValidIsoTimestamp(timestamp)) {
    errors.push({
      event_id: eventId || 'missing-id',
      error_type: 'INVALID_TIMESTAMP',
      field: 'timestamp',
      message: `Timestamp "${timestamp}" is not a valid ISO-8601 date string.`,
      raw_record: raw,
    });
  }

  // 4. Validate source
  const source = raw.source ? String(raw.source).trim().toLowerCase() : '';
  if (!source || !VALID_SOURCES.has(source)) {
    errors.push({
      event_id: eventId || 'missing-id',
      error_type: 'INVALID_SOURCE',
      field: 'source',
      message: `Source "${raw.source}" is unsupported. Supported sources: ${Array.from(VALID_SOURCES).join(', ')}.`,
      raw_record: raw,
    });
  }

  // 5. Validate event_type
  const eventType = raw.event_type ? String(raw.event_type).trim().toLowerCase() : '';
  if (!eventType || !VALID_EVENT_TYPES.has(eventType)) {
    errors.push({
      event_id: eventId || 'missing-id',
      error_type: 'INVALID_EVENT_TYPE',
      field: 'event_type',
      message: `Event type "${raw.event_type}" is unsupported. Supported types: ${Array.from(VALID_EVENT_TYPES).join(', ')}.`,
      raw_record: raw,
    });
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
}

export function validateEventBatch(rawEvents: RawEvent[]): {
  validRawEvents: RawEvent[];
  validationErrors: ValidationError[];
} {
  const validRawEvents: RawEvent[] = [];
  const validationErrors: ValidationError[] = [];

  if (!Array.isArray(rawEvents)) {
    return { validRawEvents: [], validationErrors: [{
      event_id: 'batch_error',
      error_type: 'MALFORMED_RECORD',
      field: 'payload',
      message: 'Input batch must be an array of event objects.',
    }] };
  }

  for (let i = 0; i < rawEvents.length; i++) {
    const raw = rawEvents[i];
    const { isValid, errors } = validateEvent(raw);
    if (isValid) {
      validRawEvents.push(raw);
    } else {
      validationErrors.push(...errors);
    }
  }

  return { validRawEvents, validationErrors };
}

File: src_py/validation.py
import re
from datetime import datetime

VALID_SOURCES = {
    'organic_search',
    'paid_search',
    'social',
    'email',
    'website',
    'manual'
}

VALID_EVENT_TYPES = {
    'page_visit',
    'campaign_click',
    'form_submission',
    'email_open',
    'email_click',
    'contacted',
    'qualified',
    'converted',
    'lost'
}

def is_valid_iso_timestamp(val: str) -> bool:
    if not val or not isinstance(val, str):
        return False
    # Try parsing ISO 8601 strings
    clean = val.strip().replace('Z', '+00:00')
    try:
        dt = datetime.fromisoformat(clean)
        return 1990 <= dt.year <= 2100
    except Exception:
        # Fallback date parse
        try:
            from email.utils import parsedate_to_datetime
            dt = parsedate_to_datetime(val)
            return 1990 <= dt.year <= 2100
        except Exception:
            return False

def validate_event(raw: dict):
    errors = []
    if not isinstance(raw, dict):
        return False, [{
            "event_id": "unknown",
            "error_type": "MALFORMED_RECORD",
            "field": "record",
            "message": "Event record must be a dictionary/JSON object."
        }]

    event_id = str(raw.get('event_id') or '').strip()
    if not event_id:
        errors.append({
            "event_id": "missing-id",
            "error_type": "MISSING_EVENT_ID",
            "field": "event_id",
            "message": "Event record is missing a required event_id."
        })

    email = str(raw.get('email') or '').strip()
    phone = str(raw.get('phone') or '').strip()
    lead_id = str(raw.get('lead_id') or '').strip()

    if not email and not phone and not lead_id:
        errors.append({
            "event_id": event_id or "missing-id",
            "error_type": "MISSING_LEAD_IDENTITY",
            "field": "email/phone/lead_id",
            "message": "Event must contain at least one usable identifier."
        })

    ts = raw.get('timestamp')
    if not is_valid_iso_timestamp(ts):
        errors.append({
            "event_id": event_id or "missing-id",
            "error_type": "INVALID_TIMESTAMP",
            "field": "timestamp",
            "message": f"Timestamp '{ts}' is not a valid ISO-8601 date string."
        })

    source = str(raw.get('source') or '').strip().lower()
    if not source or source not in VALID_SOURCES:
        errors.append({
            "event_id": event_id or "missing-id",
            "error_type": "INVALID_SOURCE",
            "field": "source",
            "message": f"Source '{raw.get('source')}' is unsupported."
        })

    event_type = str(raw.get('event_type') or '').strip().lower()
    if not event_type or event_type not in VALID_EVENT_TYPES:
        errors.append({
            "event_id": event_id or "missing-id",
            "error_type": "INVALID_EVENT_TYPE",
            "field": "event_type",
            "message": f"Event type '{raw.get('event_type')}' is unsupported."
        })

    return len(errors) == 0, errors

def validate_batch(raw_events: list):
    valid = []
    errors = []
    for r in raw_events:
        is_ok, errs = validate_event(r)
        if is_ok:
            valid.append(r)
        else:
            errors.extend(errs)
    return valid, errors
